crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -448,10 +475,28 @@ impl InstancePool {
         let instance = unsafe { &mut *handle.instance };
 
         // Decommit any linear memories that were used
-        for (memory, base) in instance.memories.values_mut().zip(self.memories.get(index)) {
+        for (_mem_idx, (memory, base)) in instance
+            .memories
+            .values_mut()
+            .zip(self.memories.get(index))
+            .enumerate()
+        {
             let mut memory = mem::take(memory);
             debug_assert!(memory.is_static());
 
+            #[cfg(feature = "memfd-allocator")]
+            {
+                if let Memory::Static {
+                    memfd_slot: Some(mut memfd_slot),
+                    ..
+                } = memory
+                {
+                    let _ = memfd_slot.clear_and_remain_ready();
```

Comment:
I commented below about error cases, but I think this is one error we're definitely not going to want to ignore. If the `madvise` fails for whatever reason I think that would definitely leak contents of memory between module instantiations which we don't want. On error here it seems like we should either panic to assert it's not actually happening or we discard the slot so the mappings are entirely redone on reuse.

Reply:
Oh also one thing I forgot from earlier, I feel like this wants to do the same thing as uffd ("just call madvise") but it's done in a different fashion. I think that this location (which currently madvise's manually and then `continue`s to skip further logic) would be a good place to try to better unify what uffd and memfd are doing. Ideally this could fall through to the shared call to `decommit_memory_pages` or something like that, or otherwise make it more clear that uffd and memfd are doing roughly similar things.

Reply:
As above, made this panic the process if it fails, since such a failure leaves the whole pooling allocator in an unknown state.

Re: uffd -- this gets to the bigger question of how much we can share, but: given that (i) *just* enough stuff differs in the details between what memfd does and uffd does, and (ii) uffd may go away if memfd works better, I'm somewhat hesitant to refactor to try to share more logic when it's not an easy/simple change. It starts to get into the territory of "clean up uffd", which feels out-of-scope for this PR.  My experience with trying to share too much logic as well is that sometimes it becomes hard to see through all the conditionals, and it can actually be simpler to maintain when the two disjoint approaches are kept separate. (See e.g. Cranelift backends and separate instruction lowering rules, vs. big mess of conditional logic. Not exactly the same, but similar in spirit.)